src/reports/io/hdf5_writer.cpp

```diff
@@ -0,0 +1,142 @@
+    dims[1] = Implementation::get_global_dims(report_name_, total_elements);
+    hid_t data_space = H5Screate_simple(2, dims, nullptr);
+    dataset_ = H5Dcreate(file_,
+                         dataset_name.c_str(),
+                         H5T_IEEE_F32LE,
+                         data_space,
+                         H5P_DEFAULT,
+                         H5P_DEFAULT,
+                         H5P_DEFAULT);
+
+    // Caculate the offset of each rank
+    offset_[1] = Implementation::get_offset(report_name_, total_elements);
+    H5Sclose(data_space);
+}
+
+void HDF5Writer::write(const std::vector<double>& buffer, int steps_to_write, int total_elements) {
+    hsize_t count[2];
+    count[0] = steps_to_write;
+    count[1] = total_elements;
+
+    hid_t memspace = H5Screate_simple(2, count, nullptr);
+    hid_t filespace = H5Dget_space(dataset_);
+
+    H5Sselect_hyperslab(filespace, H5S_SELECT_SET, offset_.data(), nullptr, count, nullptr);
+    /*for(int i=0; i<steps_to_write; i++) {
+        if(i==0)
+            H5Sselect_hyperslab(space, H5S_SELECT_SET, &offset_[i], NULL, &count[i], NULL);
+        H5Sselect_hyperslab(space, H5S_SELECT_OR, &offset_[i], NULL, &count[i], NULL);
+    }*/
+
+    H5Dwrite(dataset_, H5T_NATIVE_DOUBLE, memspace, filespace, H5P_DEFAULT, buffer.data());
+    offset_[0] += steps_to_write;
+
+    H5Sclose(filespace);
+    H5Sclose(memspace);
+}
+
+void HDF5Writer::write(const std::string& dataset_name, const std::vector<uint32_t>& buffer) {
+    write_any(dataset_name, buffer);
+}
+
+void HDF5Writer::write(const std::string& dataset_name, const std::vector<uint64_t>& buffer) {
+    write_any(dataset_name, buffer);
+}
+
+void HDF5Writer::write(const std::string& dataset_name, const std::vector<float>& buffer) {
+    write_any(dataset_name, buffer);
+}
+
+void HDF5Writer::write(const std::string& dataset_name, const std::vector<double>& buffer) {
+    write_any(dataset_name, buffer);
+}
+
+template <typename T>
+void HDF5Writer::write_any(const std::string& dataset_name, const std::vector<T>& buffer) {
+    hsize_t dims = buffer.size();
+    hid_t type = h5typemap::get_h5_type(T(0));
+
+    hsize_t global_dims = Implementation::get_global_dims(report_name_, dims);
+    hsize_t offset = Implementation::get_offset(report_name_, dims);
+
+    hid_t data_space = H5Screate_simple(1, &global_dims, nullptr);
+    hid_t data_set = H5Dcreate(
+        file_, dataset_name.c_str(), type, data_space, H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
+
+    hid_t filespace = H5Dget_space(data_set);
+    hid_t memspace = H5Screate_simple(1, &dims, nullptr);
+    H5Sselect_hyperslab(filespace, H5S_SELECT_SET, &offset, nullptr, &dims, nullptr);
+
+    H5Dwrite(data_set, type, memspace, filespace, collective_list_, buffer.data());
+
+    H5Sclose(memspace);
+    H5Sclose(filespace);
+    H5Dclose(data_set);
+    H5Sclose(data_space);
+}
+
+void HDF5Writer::close() {
+    // We close the dataset "/data" and the hdf5 file
+    if (dataset_) {
```

Comment:
this is a `hid_t`, does it do the right bool conversion?  What happens if someone calls `close()` twice?

Reply:
hid_t is just an int64_t
`/usr/include/hdf5/serial/H5Ipublic.h:typedef int64_t hid_t;`

If close twice you will get an hdf5 error saying that the file id is not valid

Reply:
should you set `dataset_ = 0;` after closing it then? so it won't get called again?